docs/optimistic_locking.rst

```diff
@@ -0,0 +1,167 @@
+
+  # Get a second local copy of Office
+  office_out_of_date = Office.get(office.office_id)
+
+  # Add another employee and persist the change.
+  office.employees.append(OfficeEmployeeMap(office_employee_id=str(uuid4()), person='lita'))
+  office.save()
+  # On subsequent save or update operations the version is also incremented locally to match the persisted value so
+  # there's no need to refresh between operations when reusing the local copy.
+  assert office.version == 2
+  assert office_out_of_date.version == 1
+
+The version checking is implemented using DynamoDB conditional write constraints, asserting that no value exists
+for the version attribute on the initial save and that the persisted value matches the local value on subsequent writes.
+
+
+Model.{update, save, delete}
+----------------------------
+These operations will fail if the local object is out-of-date.
+
+.. code-block:: python
+
+  @contextmanager
+  def assert_condition_check_fails():
+      try:
+          yield
+      except (PutError, UpdateError, DeleteError) as e:
+          assert isinstance(e.cause, ClientError)
+          assert e.cause_response_code == "ConditionalCheckFailedException"
+      except TransactWriteError as e:
+          assert isinstance(e.cause, ClientError)
+          assert e.cause_response_code == "TransactionCanceledException"
+          assert "ConditionalCheckFailed" in e.cause_response_message
+      else:
+          raise AssertionError("The version attribute conditional check should have failed.")
+
+
+  with assert_condition_check_fails():
+      office_out_of_date.update(actions=[Office.name.set('new office name')])
+
+  office_out_of_date.employees.remove(garrett)
+  with assert_condition_check_fails():
+      office_out_of_date.save()
+
+  # After refreshing the local copy our write operations succeed.
+  office_out_of_date.refresh()
+  office_out_of_date.employees.remove(garrett)
+  office_out_of_date.save()
+  assert office_out_of_date.version == 3
+
+  with assert_condition_check_fails():
+      office.delete()
+
+Transactions
+------------
+
+Transactions are supported.
+
+Successful
+__________
+
+.. code-block:: python
+
+  connection = Connection(host='http://localhost:8000')
+
+  office2 = Office(office_id=str(uuid4()), name="second office", employees=[justin])
+  office2.save()
+  assert office2.version == 1
+  office3 = Office(office_id=str(uuid4()), name="third office", employees=[garrett])
+  office3.save()
+  assert office3.version == 1
+
+  with TransactWrite(connection=connection) as transaction:
+      transaction.condition_check(Office, office.office_id, condition=(Office.name.exists()))
+      transaction.delete(office2)
+      transaction.save(Office(office_id=str(uuid4()), name="new office", employees=[justin, garrett]))
+      transaction.update(
+          office3,
+          actions=[
+              Office.name.set('birdistheword'),
+          ]
+      )
+
+  try:
+    office2.refresh()
```

Comment:
should we `pytest.fail` on the next line to ensure this was deleted?

Reply:
Yes, I should definitely raise an exception if `DoesNotExist` isn't thrown. I think pytest.fail would be a bit out-of-place since this isn't a test, but I'll add an else block which raises an `AssertionError`.

Reply:
Ah, didn't even realize that this wasn't a test when I was reading it 😄 Sounds good!